### DISCUSSION TITLE
Nightscout upload

### DISCUSSION THREAD

**Discussion by psonnera** [maheini/FreeStyle-Libre-3-patch#38] [Ideas]
AAPS NSClient receives data from Juggluco as an xDrip+ brodcast but cannot upload data to Nightscout. If Juggluco could do that it would integrate better with DIY solutions.
Note: using LibreView is not something I like.

**Comment by j-kaltes** [comment]
Before I could consider uploading data to Nightscout , there has to be an easy way to make use of Nightscout. I haven't found that yet.

**Reply by psonnera** [reply]
I understand you don't see this as a priority since Juggluco uploads to LV. Easy way in terms on how the API works in order to add it, or Nightscout itself?

**Comment by j-kaltes** [comment]
For users it is too difficult to set up. In the whole world at most 2000 people will use it. 

I also don't exactly understand what use it is.
I also don't need Libreview. I have the statistics screen and glucose curve in Juggluco. I also I don't see why healthcare professionals should have constant access to your glucose data. Can't you send an e-mail with some screenshots? You can also export the data and use them in Excell, R or Mathematica.
If you want someone to follow you glucose values in real-time, you can better use Juggluco's mirror function.

**Reply by duncanbarnes** [reply]
I for one do appreciate the work you've done, and I know this is a bit old, but to counter this a bit as it would be great to be able to upload to Nightscout.

> In the whole world at most 2000 people will use it.

Not sure where the 2000 number is coming from, the users are difficult to count as its predominantly a self-hosted option, but the primary Facebook group has 39,000 members and the most common (the documented) way for self-hosted installations to be deployed is using a fork, of which there are 69,000, on top of that are the users who are using a hosted service. Not all will have use for this but it's certainly considerably more than 2000.

> For users it is too difficult to set up.

It's now provided by several groups as a hosted service for those that want that option so it's 2 minutes to set up really, but there's 69,000 forks on the repository, and 172 contributors to the project so 99% of those forks are people who have deployed it.

>I also I don't see why healthcare professionals should have constant access to your glucose data. Can't you send an e-mail with some screenshots? You can also export the data and use them in Excell, R or Mathematica.

Different people have different needs and aspirations for their data certainly. My clinic for example has a large number of patients using Nightscout (alongside Diabetic clinicians). It's not about providing remote access, for many it provides the ability in the clinic to quickly and conveniently delve into data from multiple sources in a manner that screenshots would not provide. It would be a lot of work to recreate the reporting capabilities of Nightscout inside Excel, R or Mathematica, combining the data from disparate sources, especially if this is being advocated as a reporting method for users who are struggling with setting up Nightscout.

In terms of adding data to Nightscout, it has two HTTP API's which are documented and can receive entries using a straightforward POST request, the WebSockets you've picked up on are really for frontend communication.

The below query will insert a new entry into a Nightscout installation running on 127.0.0.1:1337

curl -X POST -H 'api-secret: c44b91d6672bb1d7c51e050ababb0146efa84380' -H 'accept:  */*' -H 'Content-Type: application/json' -i 'http://127.0.0.1:1337/api/v1/entries.json' --data '[{"type": "sgv", "date": 1682808021000, "sgv": 111, "device":"Juggluco"}]'

The api-secret parameter is a SHA1 hash of the actual value, in the above example the unencrypted Nightscout API secret is: 1234567890abc

Additional parameters which can be added to the query include:

- delta - INT - Trend direction
- direction - STRING - Trend direction string, e.g. Flat, FortyFiveUp, FortyFiveDown

Hope that's of some help!

**Reply by maheini** [reply]
Well, it does seem like jkaltes did already implemented this feature in the newer versions... just navigate to Juggluco -> Settings -> Uploader and enter your url & secret

**Reply by duncanbarnes** [reply]
Right ok, didn't see just found it is on GitHub as well, didn't think it was and assumed this was it. To be honest the lack of this is one of the main reasons I stopped using it so will give it another try now if that works.

**Comment by j-kaltes** [comment]
I have looked at it. The Nightscout server for xDrip, Nightwatch and Garmin watches differs little from the xDrip webserver Juggluco already incorporates.
If you turn on xDrip webserver in Juggluco, you can just fill in http://127.0.0.1:17580 for the Nightscout server and xDrip receives the values from Juggluco.
For Nightwatch and Garmin watches you have only to get a SSL key that is authenticated (for example for free from https://zerossl.com).

But Nsclient seems to be impossible. What the hell do I have to make of:

```
GET /socket.io/?EIO=3&transport=polling HTTP/1.1

96:0{"sid":"mE7M1PSeHBvTxbRlAAAG","upgrades":["websocket"],"pingInterval":25000,"pingTimeout":5000}2:40

POST /socket.io/?EIO=3&transport=polling&sid=mE7M1PSeHBvTxbRlAAAG HTTP/1.1

GET /socket.io/?EIO=3&transport=websocket&sid=mE7M1PSeHBvTxbRlAAAG HTTP/1.1

```
When I grep for "socket.io" in the source of AndroidAPS, I can only find the following:
```
core/core_dependencies.gradle:    api("io.socket:socket.io-client:1.0.0") {
.git/packed-refs:e21e93057bab9ad78c8e1195cff6aa518c70167d refs/remotes/origin/dependabot/gradle/dev/io.socket-socket.io-client-2.1.0
```
I could find something io.socket.engineio.client.transports.Polling. Probably you have to use io.socket yourself to be able to talk to it.

**Comment by Suchiman** [comment]
@j-kaltes that is socket.io which is based on engine.io to provide a real time eventing communication channel. AFAIK it is on its way out, AAPS is migrating to the v3 REST API of nightscout https://github.com/nightscout/cgm-remote-monitor/blob/master/lib/api3/doc/tutorial.md#create which can be trivially used by any HTTP Client

**Comment by psonnera** [comment]
Thank you @j-kaltes for the implementation. I now need to understand port mapping for PAAS (Fly, Railway, etc...) and others to add Juggluco to NS docs.
